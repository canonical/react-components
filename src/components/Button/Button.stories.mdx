import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import Button from "./Button";

<Meta
  title="Button"
  component={Button}
  argTypes={{
    children: {
      control: {
        type: "text",
      },
      defaultValue: "Click me!",
    },
  }}
/>

export const Template = (args) => <Button {...args} />;

### Button

This is a [React](https://reactjs.org/) component for the Vanilla [Button](https://docs.vanillaframework.io/patterns/buttons/).

Buttons are clickable elements used to perform an action, they can be used for buttons and link elements.

### Props

<ArgsTable of={Button} />

### Default

<Canvas>
  <Story
    name="Default"
    args={{
      children: "Button",
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<Canvas>
  <Story
    name="Default disabled"
    args={{
      children: "Button",
      disabled: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Base

<Canvas>
  <Story name="Base">
    <Button appearance="base">Base button</Button>
  </Story>
</Canvas>

<Canvas>
  <Story name="Base disabled">
    <Button appearance="base" disabled>
      Base button disabled
    </Button>
  </Story>
</Canvas>

### Link

<Canvas>
  <Story name="Link">
    <Button element="a" href="#test" appearance="base">
      Link button
    </Button>
  </Story>
</Canvas>

<Canvas>
  <Story name="Link disabled">
    <Button element="a" href="#test" appearance="base" disabled>
      Link button disabled
    </Button>
  </Story>
</Canvas>

### Positive

<Canvas>
  <Story name="Positive">
    <Button appearance="positive">Positive button</Button>
  </Story>
</Canvas>

<Canvas>
  <Story name="Positive disabled">
    <Button appearance="positive" disabled>
      Positive button disabled
    </Button>
  </Story>
</Canvas>

### Negative

<Canvas>
  <Story name="Negative">
    <Button appearance="negative">Negative button</Button>
  </Story>
</Canvas>

<Canvas>
  <Story name="Negative disabled">
    <Button appearance="negative" disabled>
      Negative button disabled
    </Button>
  </Story>
</Canvas>

### Brand

<Canvas>
  <Story name="Brand">
    <Button appearance="brand">Brand button</Button>
  </Story>
</Canvas>

<Canvas>
  <Story name="Brand disabled">
    <Button appearance="brand" disabled>
      Brand button disabled
    </Button>
  </Story>
</Canvas>

### Inline

<Canvas>
  <Story name="Inline">
    <>
      <span>Everything you need to get started with Vanilla. </span>
      <Button appearance="neutral" inline>
        Inline button
      </Button>
    </>
  </Story>
</Canvas>

### Dense

<Canvas>
  <Story name="Dense">
    <>
      <span>Everything you need to get started with Vanilla. </span>
      <Button dense>Dense button</Button>
    </>
  </Story>
</Canvas>

### Small

<Canvas>
  <Story name="Small">
    <>
      <Button small>Small button</Button>
      <Button small dense>
        Small dense button
      </Button>
    </>
  </Story>
</Canvas>

### Icon

<Canvas>
  <Story name="Icon">
    <Button hasIcon>
      <i className="p-icon--plus"></i>
    </Button>
  </Story>
</Canvas>

<Canvas>
  <Story name="Icon & text">
    <Button hasIcon>
      <i className="p-icon--plus"></i> <span>Button with icon & text</span>
    </Button>
  </Story>
</Canvas>
