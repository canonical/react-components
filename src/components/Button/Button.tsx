import classNames from "classnames";
import React from "react";
import type {
  ButtonHTMLAttributes,
  ComponentType,
  ElementType,
  MouseEventHandler,
  ReactNode,
} from "react";

import type { ClassName, ValueOf } from "types";

export const ButtonAppearance = {
  BASE: "base",
  BRAND: "brand",
  DEFAULT: "",
  LINK: "link",
  NEGATIVE: "negative",
  POSITIVE: "positive",
} as const;

/**
 * The type of the Button props.
 * @template P - The type of the props if providing a component to `element`
 */
export type Props<P = null> = {
  /**
   * The appearance of the button.
   */
  appearance?: ValueOf<typeof ButtonAppearance> | string;
  /**
   * The content of the button.
   */
  children?: ReactNode;
  /**
   * Optional class(es) to pass to the button element.
   */
  className?: ClassName;
  /**
   * Whether the button should have dense padding.
   */
  dense?: boolean;
  /**
   * Whether the button should be disabled.
   */
  disabled?: boolean;
  /**
   * Optional element or component to use instead of `button`.
   */
  element?: ElementType | ComponentType<P>;
  /**
   * Whether the button has an icon in the content.
   */
  hasIcon?: boolean;
  /**
   * Whether the button should display inline.
   */
  inline?: boolean;
  /**
   * Function for handling button click event.
   */
  onClick?: MouseEventHandler<HTMLButtonElement>;
  /**
   * Whether the button should be small.
   */
  small?: boolean;
} & (Omit<ButtonHTMLAttributes<HTMLButtonElement>, "onClick"> | P);

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [Button](https://docs.vanillaframework.io/patterns/buttons/).
 *
 * Buttons are clickable elements used to perform an action, they can be used for buttons and link elements.
 * @template P - The type of the props if providing a component to `element`
 */
const Button = <P,>({
  appearance,
  children,
  className,
  dense,
  disabled,
  element: Component = "button",
  hasIcon,
  inline,
  onClick,
  small,
  ...buttonProps
}: Props<P>): JSX.Element => {
  const classes = classNames(
    appearance ? `p-button--${appearance}` : "p-button",
    {
      "has-icon": hasIcon,
      "is-dense": dense,
      "is-disabled": disabled,
      "is-inline": inline,
      "is-small": small,
    },
    className
  );
  const onClickDisabled = (e: MouseEvent) => e.preventDefault();

  return (
    <Component
      className={classes}
      onClick={disabled ? onClickDisabled : onClick}
      aria-disabled={disabled || undefined}
      {...buttonProps}
    >
      {children}
    </Component>
  );
};

export default Button;
