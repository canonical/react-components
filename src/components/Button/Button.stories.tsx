import React, { HTMLProps } from "react";
import { Meta, StoryObj } from "@storybook/react";

import Tooltip from "components/Tooltip";

import Button, { ButtonAppearance } from "./Button";

const meta: Meta<typeof Button> = {
  component: Button,
  tags: ["autodocs"],

  argTypes: {
    children: {
      control: {
        type: "text",
      },
    },

    appearance: {
      control: {
        options: ButtonAppearance,
        type: "select",
      },
    },
  },

  args: {
    children: "Click me!",
    appearance: ButtonAppearance.DEFAULT,
  },
};

export default meta;

type Story<P = null> = StoryObj<typeof Button<P>>;

export const Default: Story = {
  name: "Default",

  args: {
    children: "Button",
  },
};

export const DefaultDisabled: Story = {
  name: "Default disabled",

  args: {
    children: "Button",
    disabled: true,
  },
};

export const Base: Story = {
  name: "Base",

  args: {
    children: "Base button",
    appearance: "base",
  },
};

export const BaseDisabled: Story = {
  name: "Base disabled",

  args: {
    children: "Base button disabled",
    appearance: "base",
    disabled: true,
  },
};

export const Link = {
  name: "Link",

  args: {
    children: "Link button",
    appearance: "base",
    element: "a",
    href: "#test",
  },
};

export const LinkDisabled: Story<HTMLProps<HTMLAnchorElement>> = {
  name: "Link disabled",

  args: {
    children: "Link button disabled",
    appearance: "base",
    element: "a",
    href: "#test",
    disabled: true,
  },
};

export const Positive: Story = {
  name: "Positive",

  args: {
    children: "Positive button",
    appearance: "positive",
  },
};

export const PositiveDisabled: Story = {
  name: "Positive disabled",

  args: {
    children: "Positive button disabled",
    appearance: "positive",
    disabled: true,
  },
};

export const Negative: Story = {
  name: "Negative",

  args: {
    children: "Negative button",
    appearance: "negative",
  },
};

export const NegativeDisabled: Story = {
  name: "Negative disabled",

  args: {
    children: "Negative button disabled",
    appearance: "negative",
    disabled: true,
  },
};

export const Brand: Story = {
  name: "Brand",

  args: {
    children: "Brand button",
    appearance: "brand",
  },
};

export const BrandDisabled: Story = {
  name: "Brand disabled",

  args: {
    children: "Brand button disabled",
    appearance: "brand",
    disabled: true,
  },
};

export const Inline: Story = {
  render: () => (
    <>
      <span>Everything you need to get started with Vanilla.</span>
      <Button appearance="neutral" inline>
        Inline button
      </Button>
    </>
  ),

  name: "Inline",
};

export const Dense: Story = {
  render: () => (
    <>
      <span>Everything you need to get started with Vanilla.</span>
      <Button dense>Dense button</Button>
    </>
  ),

  name: "Dense",
};

export const Small: Story = {
  render: () => (
    <>
      <Button small>Small button</Button>
      <Button small dense>
        Small dense button
      </Button>
    </>
  ),

  name: "Small",
};

export const Icon: Story = {
  name: "Icon",

  args: {
    children: <i className="p-icon--plus"></i>,
    hasIcon: true,
  },
};

export const IconText: Story = {
  name: "Icon & text",

  args: {
    children: (
      <>
        <i className="p-icon--plus"></i> <span>Button with icon & text</span>
      </>
    ),

    hasIcon: true,
  },
};

export const IconWithTooltip: Story = {
  name: "Disabled with tooltip",

  args: {
    children: "Disabled button with a tooltip",
    disabled: true,
  },
  render: (args) => (
    <div style={{ paddingTop: "3rem" }}>
      <Tooltip message="This button is disabled">
        <Button {...args} />
      </Tooltip>
    </div>
  ),
};
