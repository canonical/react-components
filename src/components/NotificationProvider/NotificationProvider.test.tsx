import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import {
  NotificationConsumer,
  NotificationProvider,
  useNotify,
} from "./NotificationProvider";
import Button from "../Button";
import { act } from "react-dom/test-utils";

describe("NotificationProvider", () => {
  it("stores and renders notifications", async () => {
    const handleRetry = jest.fn();

    const TriggerComponent = () => {
      const notify = useNotify();
      const handleSuccess = () => notify.success("My success!");
      const handleClear = () => notify.clear();
      const handleInfo = () => notify.info("Some more details", "Test info");
      const handleError = () =>
        notify.failure(
          "Fail title",
          new Error("error message"),
          "Button caused a failure",
          [
            {
              label: "Retry",
              onClick: handleRetry,
            },
          ]
        );

      return (
        <div>
          <Button onClick={handleSuccess} data-testid="success-btn" />
          <Button onClick={handleClear} data-testid="clear-btn" />
          <Button onClick={handleError} data-testid="error-btn" />
          <Button onClick={handleInfo} data-testid="info-btn" />
        </div>
      );
    };

    render(
      <div>
        <NotificationProvider>
          <TriggerComponent />
          <div data-testid="notification-consumer">
            <NotificationConsumer />
          </div>
        </NotificationProvider>
      </div>
    );

    const clickBtn = async (testId: string) =>
      await act(async () => await userEvent.click(screen.getByTestId(testId)));

    expect(screen.getByTestId("notification-consumer")).toBeEmptyDOMElement();

    await clickBtn("success-btn");
    expect(screen.getByTestId("notification-consumer")).toMatchSnapshot();

    await clickBtn("clear-btn");
    expect(screen.getByTestId("notification-consumer")).toBeEmptyDOMElement();

    await clickBtn("error-btn");
    expect(screen.getByTestId("notification-consumer")).toMatchSnapshot();

    expect(handleRetry).not.toHaveBeenCalled();
    await clickBtn("notification-action");
    expect(handleRetry).toHaveBeenCalled();

    await clickBtn("info-btn");
    expect(screen.getByTestId("notification-consumer")).toMatchSnapshot();
  });
});
