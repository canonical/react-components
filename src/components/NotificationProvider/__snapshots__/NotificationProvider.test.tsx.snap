// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationProvider stores and renders notifications 1`] = `
<div
  data-testid="notification-consumer"
>
  <div>
    <div
      class="p-notification--positive"
    >
      <div
        class="p-notification__content"
      >
        <h5
          class="p-notification__title"
          data-testid="notification-title"
        >
          Test success
        </h5>
        <p
          class="p-notification__message"
        >
          My success!
        </p>
        <button
          class="p-notification__close"
          data-testid="notification-close-button"
          type="button"
        >
          Close notification
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`NotificationProvider stores and renders notifications 2`] = `
<div
  data-testid="notification-consumer"
>
  <div>
    <div
      class="p-notification--negative"
    >
      <div
        class="p-notification__content"
      >
        <h5
          class="p-notification__title"
          data-testid="notification-title"
        >
          Fail title
        </h5>
        <p
          class="p-notification__message"
        >
          Button caused a failure
           
          error message
        </p>
        <button
          class="p-notification__close"
          data-testid="notification-close-button"
          type="button"
        >
          Close notification
        </button>
      </div>
      <div
        class="p-notification__meta"
        data-testid="notification-meta"
      >
        <div
          class="p-notification__actions"
        >
          <button
            class="p-button--link p-notification__action"
            data-testid="notification-action"
            type="button"
          >
            Retry
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`NotificationProvider stores and renders notifications 3`] = `
<div
  data-testid="notification-consumer"
>
  <div>
    <div
      class="p-notification--information"
    >
      <div
        class="p-notification__content"
      >
        <h5
          class="p-notification__title"
          data-testid="notification-title"
        >
          Test info
        </h5>
        <p
          class="p-notification__message"
        >
          Some more details
        </p>
        <button
          class="p-notification__close"
          data-testid="notification-close-button"
          type="button"
        >
          Close notification
        </button>
      </div>
    </div>
  </div>
</div>
`;
